src/Workspaces/Core/Portable/FindSymbols/FindReferences/FindReferencesSearchEngine.cs

```diff
@@ -281,39 +297,37 @@ private async ValueTask<SymbolGroup> ReportGroupAsync(ISymbol symbol, Cancellati
         // Note: cascaded symbols will normally have the same name.  That's ok.  The second call to
         // FindMatchingIdentifierTokens with the same name will short circuit since it will already see the result of
         // the prior call.
-        foreach (var symbol in symbols)
+        foreach (var (symbol, _) in symbolToSymbolGroup)
         {
             if (symbol.CanBeReferencedByName)
                 cache.FindMatchingIdentifierTokens(symbol.Name, cancellationToken);
         }
 
         await RoslynParallel.ForEachAsync(
-            symbols,
+            symbolToSymbolGroup,
             GetParallelOptions(cancellationToken),
-            (symbol, cancellationToken) =>
+            (kvp, cancellationToken) =>
             {
+                var (symbol, group) = kvp;
+
                 // symbolToGlobalAliases is safe to read in parallel.  It is created fully before this point and is no
                 // longer mutated.
                 var state = new FindReferencesDocumentState(
                     cache, TryGet(symbolToGlobalAliases, symbol));
 
-                ProcessDocument(symbol, state, onReferenceFound);
+                ProcessDocument(symbol, group, state, onReferenceFound);
                 return ValueTaskFactory.CompletedTask;
             }).ConfigureAwait(false);
 
         return;
 
         void ProcessDocument(
-            ISymbol symbol, FindReferencesDocumentState state, Action<Reference> onReferenceFound)
+            ISymbol symbol, SymbolGroup group, FindReferencesDocumentState state, Action<Reference> onReferenceFound)
         {
             cancellationToken.ThrowIfCancellationRequested();
 
             using (Logger.LogBlock(FunctionId.FindReference_ProcessDocumentAsync, cancellationToken))
             {
-                // This is safe to just blindly read. We can only ever get here after the call to ReportGroupsAsync
-                // happened.  So there must be a group for this symbol in our map.
-                var group = _symbolToGroup[symbol];
```

Comment:
the crux of hte change is that instead of having a collection be blindly written and read from from disparate locations, we just take t6he point that we write the data initially and pass that data along all the way to here so we're never out of sync.